HTTP API support

cretz opened the pull request:
**What changed?**

Implements support for HTTP API as proposed at https://github.com/temporalio/proposals/pull/79. Specifically:

* Added `httpPort` ~and `httpDisabled`~ and `httpAdditionalForwardedHeaders` on config RPC
  * Only applicable to frontend service
  * ~Default port is gRPC port + 10, default `httpDisabled` is `false` (meaning this is on by default)~
  * Default of port 0 now means HTTP disabled but all docker configs set to default to 7243
  * There is a default set of headers as visible from code
  * This config should be discussed to see if we're ok with it
* Added `HTTPAPIServer` for frontend which runs HTTP API server using gRPC gateway with our custom modifications which include
  * Ability to support `?pretty` to pretty print JSON and/or `?noPayloadShorthand` to disable shorthand payloads
  * Uses direct gRPC server invocation via invocation of gRPC interceptors
  * Implements well-defined graceful stop and adds basic common Go HTTP server protection measures
  * Custom error handler due to gogo status vs Google gRPC status issues with `Any` details
  * Properly forwards headers and TLS auth info so claim mapper and authorizer works
* Added `HTTPAPIServerProvider` for frontend which creates `HTTPAPIServer`
* Updated frontend `Service` to support `HTTPAPIServer`
  * Make gRPC server and HTTP server start concurrently
  * Make gRPC server and HTTP server shutdown concurrently
* Added tests for HTTP API
* Added basic mTLS integration tests for gRPC and HTTP API

This is ready for review. This will not pass CI until the other PRs are merged. Related PRs:

* Depends on https://github.com/temporalio/api/pull/301
* Depends on https://github.com/temporalio/api-go/pull/112

cretz commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,221 @@
+	"go.temporal.io/server/common/util"
+)
+
+type HTTPAPIServer struct {
```
Need Godoc

cretz replied:
Done

cretz commented on tests/http_api_test.go:
```diff
@@ -0,0 +1,148 @@
+		"workflowType": { "name": "http-basic-workflow" },
+		"taskQueue": { "name": "`+s.taskQueue+`" },
+		"input": [{ "someField": "workflow-arg" }],
+		"requestId": "`+s.randomizeStr("req")+`"
```
I may, in this PR or separately if approved by server team, change `requestId` from being required to being optional

cretz replied:
Done

cretz commented on tests/http_api_test.go:
```diff
@@ -0,0 +1,148 @@
+	}`)
+	var errResp struct {
+		Message string `json:"message"`
+		Details []struct {
+			RunID string `json:"runId"`
+		} `json:"details"`
```
For this style of details to work, I had to implement a custom error handler because the default reusing our marshaler doesn't work with Google "Any" wkt, and we want to use our marshaler.

cretz commented on tests/http_api_test.go:
```diff
@@ -0,0 +1,148 @@
+	_, respBody = s.httpPost(
+		http.StatusOK,
+		"/api/v1/namespaces/"+s.namespace+"/workflows/"+workflowID+"/query/some-query",
+		`{ "query": { "queryArgs": [{ "someField": "query-arg" }] } }`,
```
Notice the inconsistencies in our API come to the forefront here. What is `query.queryArgs` here is `input` for signal/workflow.

MichaelSnowden replied:
These are new APIs, though. Why not address this?

cretz replied:
You cannot address this reasonably with gRPC gateway. Our HTTP API, as a rule and as a byproduct of our approach, exposes the RPCs we have made, it does not alter then. We chose to do this in our API, gRPC and HTTP users have to deal with it. This is covered a bit better in the proposal: https://github.com/temporalio/proposals/blob/master/api/http-api.md.

If we want to fix our RPCs to be more consistent, HTTP and gRPC will benefit. If not, HTTP and gRPC calls will be inconsistent across calls (but consistent across protocols).

cretz commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,221 @@
+	}
+
+	// Create a client to frontend gRPC
+	frontendConn := rpcFactory.CreateLocalFrontendGRPCConnection()
```
> And goes through TLS again. If you want it to stay in-process you'll need to do something different.

Yeah, I think you may be right. I am going to have the gRPC server probably listen on another listener. I don't want to use a unix socket because of platform specific concerns, I wonder if we have an in-process/memory conn/listener we use elsewhere in this repo.

> Also (at first glance) I think this bypasses authorization.. the grpc connection will look like an internal connection.

Ah I was not aware the auth checkers distinguished the two. I will need to find where that's done and make sure my local/pipe is not seen as authenticated.

> In fact if the http server terminates TLS itself (which I suppose is required) I think getting this to handle tls-based authn correctly will be a bunch of work

I'll see if I can easily do what infra does and put the client cert in an internal trusted header that I then pull back off in the mTLS authorizer. As for JWT, I figured just forwarding the headers would be good enough.

> I suppose you could stuff it in the context.

Yeah basically that.

Thanks for the feedback! I figured I needed to make these changes, I just didn't get to the tests yet.

dnr replied:
> Also, quick glance shows gRPC server doesn't support multiple serve calls, so I'm gonna have to have some kind of multi listener.

:(  I ran into a bunch of these limitations with the internal-frontend work and was mostly able to avoid it but I think you may have to deal with it now.

Crazy idea: grab a reference to the service and to the interceptor stack and implement interception manually? Might be easier + more efficient than messing with listeners and redoing serialization.

> I'll see if I can easily do what infra does and put the client cert in an internal trusted header (or context) that I then pull back off in the mTLS authorizer. As for JWT, I figured just forwarding the headers would be good enough.

It's not in the authorizer, it's in the authorization interceptor. And yeah, that should work for JWT. For mTLS... you can make the authorization interceptor pull out `credentials.AuthInfo` from somewhere else, but the bigger problem is that authorization plugins (which we can't control) get a context and can call `peer.FromContext` themselves (this is gross but...), and I assume you can't replace that context value since grpc will set it. Unless you do the crazy idea and bypass most of the stack and call the handler directly.

Feel free to ping me on slack, I'm (sadly) pretty familiar with this area

cretz replied:
> Crazy idea: grab a reference to the service and to the interceptor stack and implement interception manually? Might be easier + more efficient than messing with listeners and redoing serialization.

I'm concerned I won't get all the details right

> Feel free to ping me on slack, I'm (sadly) pretty familiar with this area

:+1: Let me get some tests going around mTLS and see what my options are here

cretz replied:
Ok, I updated the code to support inline gRPC invocation via direct interceptor invocation. Going to work on auth support/tests next.

cretz replied:
Done

cretz commented on go.mod:
```diff
@@ -148,3 +149,5 @@ require (
 	modernc.org/token v1.1.0 // indirect
 )
+
+replace go.temporal.io/api => ../temporal-api-go
```
This is of course temporary pending merge of dependencies

MichaelSnowden commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+) (*HTTPAPIServer, error) {
+	// Create a TCP listener the same as the frontend one but with different port
+	tcpAddrRef, _ := grpcListener.Addr().(*net.TCPAddr)
+	if tcpAddrRef == nil {
+		return nil, fmt.Errorf("must use TCP for gRPC listener to support HTTP API")
+	}
```
why not just use `ok`?

MichaelSnowden replied:
also, I'd include a %T somewhere

cretz replied:
Don't need `ok`, why use it? I don't think we need Go types on this error message.

MichaelSnowden commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+	tcpAddr := *tcpAddrRef
+	tcpAddr.Port = rpcConfig.HTTPPort
+	if tcpAddr.Port == 0 {
+		tcpAddr.Port = tcpAddrRef.Port + 10
```
Adding 10 here to the RPC port might work for our static configs, but it could conflict with others--it's also just too much implicit behavior IMO.

cretz replied:
Are you suggesting I open on `7243` by default regardless of your gRPC port? I can do that, though would like others' opinions.

dnr replied:
The +10 feels too magical and surprising to me also. I'd rather just require an explicit port here. All port selection should be done in static config (or by something wrapping the server and generating a static config in memory)

Feel free to add an explicit port to all the development configs and dev servers

cretz replied:
Do we require an explicit port for gRPC today? What does the absence of a RPC port in config mean today?

cretz replied:
Great! That's information I didn't know. Yes, I will make this required and set the default in static configs. I assume this is only `docker/config_template.yaml`? Can you confirm for me what happens if frontend gRPC port config is left off? (I can try to dig as well, but I assume it picks a random port which makes it not very usable)

cretz replied:
This has been fixed to require an HTTP port since we require it for gRPC. The absence of an HTTP port (or 0) means disabled. Docker config templates have been updated to set to 7243 by default.

MichaelSnowden commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+	}
+	tcpAddr := *tcpAddrRef
+	tcpAddr.Port = rpcConfig.HTTPPort
+	if tcpAddr.Port == 0 {
```
It should be ok to specify the zero port and get a random available port.

cretz replied:
Do we support that for gRPC port today? How do you differentiate an unset YAML int from 0?

MichaelSnowden commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+
+// GracefulStop stops the HTTP server. This will first attempt a graceful stop
+// with a drain time, then will hard-stop. This will not return until stopped.
+func (h *HTTPAPIServer) GracefulStop() {
```
This should accept a context because we have one at the top-level from fx (although you may have to plumb it down from there)

cretz replied:
We don't use the context for bounding gRPC server graceful stop, why is it a new requirement for HTTP server graceful stop? I intentionally chose to match what gRPC server does.

MichaelSnowden commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+	err := h.server.Serve(h.listener)
+	// If the error is for close, we have to wait for the shutdown to complete and
+	// we don't consider it an error
+	if err == http.ErrServerClosed {
```
Please use errors.Is

cretz replied:
No, I don't want to check whether it is anywhere in the hierarchy. The docs say that it returns exactly this, I check exactly this on purpose.

cretz replied:
Was required to change this unfortunately by the linter

MichaelSnowden commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+func (h *HTTPAPIServer) GracefulStop() {
+	// We try a graceful stop for the amount of time we can drain, then we do a
+	// hard stop
+	shutdownCtx, _ := context.WithTimeout(context.Background(), h.shutdownDrainTime)
```
why ignore the cancel?

cretz replied:
Because we hard-stop which gives us the guarantee this does. I can add defer cancel, it'd be effectively a no-op with how docs say `Close` works.

dnr replied:
Doesn't ignoring the cancel leak a goroutine (sometimes)?

cretz replied:
Not sure it's a goroutine specifically, but Godoc is clear that I should be cancelling here to release resources. I will add a `defer cancel` here.

cretz replied:
Fixed

dnr commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+		return nil, fmt.Errorf("failed listening for HTTP API on %v: %w", &tcpAddr, err)
+	}
+	// Close the listener if anything else in this function fails
+	success := false
```
used named return values and check that instead of an extra local var?

cretz replied:
I usually choose this approach because users can get confused on when named return values are set and especially if you name it `err` it can often be swallowed in confusing ways. But I will change if preferred.

dnr commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+	if _, ok := r.URL.Query()["noPayloadShorthand"]; ok {
+		acceptHeaderSuffix += "+no-payload-shorthand"
+	}
+	if acceptHeaderSuffix != "" {
```
What if Accept is set to something other than application/json to start? we just override it? Do we support proto over http? Or is the idea that if you want to use proto you can use grpc and http is only for json?

cretz replied:
This was mostly taken from the approach at https://grpc-ecosystem.github.io/grpc-gateway/docs/mapping/customizing_your_gateway/#pretty-print-json-responses-when-queried-with-pretty. I believe we want to support only `application/json` at this time and assume the absence of an accept header means that.

What I will do is fail at the top of this call if the `Accept` header is present and has any value other than `application/json`. That seem reasonable?

cretz replied:
Updated

dnr commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+		if conn, _ := r.Context().Value(httpRemoteAddrContextKey{}).(net.Conn); conn != nil {
+			addr = conn.RemoteAddr()
+		}
+		r = r.WithContext(peer.NewContext(r.Context(), &peer.Peer{
```
nice. this means we don't have to change the authorization interceptor at all?

cretz replied:
Exactly. And I have tests proving it.

dnr commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+	if merr != nil {
+		h.logger.Warn("Failed to marshal error message", tag.Error(merr))
+		w.WriteHeader(http.StatusInternalServerError)
+		_, _ = w.Write([]byte(`{"code": 13, "message": "failed to marshal error message"}`))
```
should set content-type to json here? or is `marshaler.ContentType()` always json?

cretz replied:
`marshaler.ContentType()` is always JSON at this time, but I'll add it explicitly just in case. I expect this code will never be hit.

cretz replied:
Fixed

dnr commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+			if !validRPCMethod {
+				continue
+			}
+			fullMethod := "/" + qualifiedServerName + "/" + serverVal.Type().Method(i).Name
```
these indexes are guaranteed to line up? it looks like yes though the reflect docs are not 100% clear

you could loop over the type methods and look up the value method by name?

cretz replied:
They are because `value.NumMethod()` is basically `value.Type().NumMethod()`, see https://github.com/golang/go/blob/449ef3795d8054faf4a601d8d1aab1f624b822f5/src/reflect/value.go#L2037. But yes, no guarantee. I can change.

cretz replied:
Fixed

dnr commented on service/frontend/service.go:
```diff
@@ -345,10 +349,22 @@ func (s *Service) Start() {
-	if err := s.server.Serve(listener); err != nil {
-		logger.Fatal("Failed to serve on frontend listener", tag.Error(err))
+
+	// Start the gRPC and HTTP server (if any) in the background
```
I'm changing this in https://github.com/temporalio/temporal/pull/4584, so after that lands, you can just start two goroutines that check return value independently, and not have to do the channel here

cretz replied:
I see. If that lands first I will change when I conflict.

dnr replied:
that one landed

cretz replied:
Will fix

dnr commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,438 @@
+	h := &HTTPAPIServer{
+		listener:          listener,
+		logger:            logger,
+		shutdownDrainTime: util.Max(time.Second, serviceConfig.ShutdownDrainDuration()),
```
I would rather pass the drain duration in to GracefulStop, instead of reading the config from two places (at two different times so they could differ). I don't really care that the signature of GracefulStop would differ (the semantics already differ in that the grpc one has no implicit timeout)

cretz replied:
:+1: I can do this

cretz replied:
Done :+1:

cretz commented:
Added `http_service_requests` metric, made `StartWorkflowExecutionRequest.request_id` optional, and added general PR updates. I won't be constantly fixing conflicts from `master` until this is nearer approval.

cretz commented on service/frontend/fx.go:
```diff
@@ -599,6 +611,38 @@ func HandlerProvider(
+	logger log.Logger,
+) (*HTTPAPIServer, error) {
+	// If HTTP API server not enabled, return nil
+	rpcConfig := config.Services[string(serviceName)].RPC
```
:+1: I was not aware there was an internal-frontend reusing this same package/code. Will only apply conditional to ensure only the public frontend gets the HTTP API.

cretz replied:
Done

cretz commented on service/frontend/http_api_server.go:
```diff
@@ -0,0 +1,472 @@
+	// moment using gRPC's default at
+	// https://github.com/grpc/grpc-go/blob/0673105ebcb956e8bf50b96e28209ab7845a65ad/server.go#L58.
+	// Max header bytes is defaulted by Go to 1MB.
+	r.Body = http.MaxBytesReader(w, r.Body, 1024*1024*4)
```
Will do. Will call it `maxHttpRequestBytes` for now, but would expect it to be updated to share with the gRPC one if/when y'all get around to it.

dnr commented on docker/config_template.yaml:
```diff
@@ -304,6 +304,7 @@ services:
             grpcPort: {{ $temporalGrpcPort }}
             membershipPort: {{ default .Env.FRONTEND_MEMBERSHIP_PORT "6933" }}
             bindOnIP: {{ default .Env.BIND_ON_IP "127.0.0.1" }}
+            httpPort: {{ default .Env.FRONTEND_HTTP_PORT "7243" }}
```
You might want to check if there are any changes to be made here:
https://github.com/search?q=repo%3Atemporalio%2Fdocker-compose%207233&type=code
or here:
https://github.com/search?q=repo%3Atemporalio%2Fhelm-charts%207233&type=code
to expose the port

cretz replied:
:+1: Will open PRs to those upon merge here.

dnr commented on service/frontend/workflow_handler.go:
```diff
@@ -369,7 +369,9 @@ func (wh *WorkflowHandler) StartWorkflowExecution(ctx context.Context, request *
-		return nil, errRequestIDNotSet
+		// For easy direct API use, we default the request ID here but expect all
+		// SDKs and other auto-retrying clients to set it
+		request.RequestId = uuid.New()
```
Do you think it's worth doing this only for http?

cretz replied:
Doing any call-specific things for HTTP is not only a bit difficult, it violates the spirit of "generic HTTP interface over existing gRPC behavior". So far I have been able to avoid any call-specific HTTP behavior.

feedmeapples commented on go.mod:
```diff
@@ -146,3 +147,5 @@ require (
 	modernc.org/token v1.1.0 // indirect
 )
+
+replace go.temporal.io/api => github.com/feedmeapples/api-go v1.24.0
```
```suggestion
```

cretz replied:
Done
